### PR TITLE
Reduce memory allocations and speed up text localization

### PR DESCRIPTION
**Pull request by nzdev**
### Prerequisites

- [x] I have added steps to test this contribution in the description below

If there's an existing issue for this PR then this fixes N/A

### Description
What?
Greatly reduce memory allocations and speed up text localisation by optimising TextLocalizationService.

Why?
Memory profiling showed megabytes of allocations caused by TextLocalizationService.Localize() while just visiting nodes in the backoffice.

```
// * Summary *

BenchmarkDotNet=v0.11.3, OS=Windows 10.0.18362
Intel Core i5-8265U CPU 1.60GHz (Kaby Lake R), 1 CPU, 8 logical and 4 physical cores
  [Host]     : .NET Framework 4.7.2 (CLR 4.0.30319.42000), 32bit LegacyJIT-v4.8.4250.0
  Job-JIATTD : .NET Framework 4.7.2 (CLR 4.0.30319.42000), 32bit LegacyJIT-v4.8.4250.0

IterationCount=3  IterationTime=100.0000 ms  LaunchCount=1
WarmupCount=3

                     Method |       Mean |      Error |    StdDev | Ratio | RatioSD | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
--------------------------- |-----------:|-----------:|----------:|------:|--------:|------------:|------------:|------------:|--------------------:|
   OriginalDictionaryGetAll |  23.015 ms | 44.5607 ms | 2.4425 ms |  0.21 |    0.01 |   1750.0000 |           - |           - |           6008954 B |
          OriginalXmlGetAll | 101.288 ms | 84.2304 ms | 4.6170 ms |  0.94 |    0.05 |  13000.0000 |           - |           - |          43467420 B |
 OriginalDictionaryLocalize |  17.674 ms | 17.5819 ms | 0.9637 ms |  0.16 |    0.00 |   1000.0000 |           - |           - |           3766319 B |
        OriginalXmlLocalize | 108.289 ms | 83.5021 ms | 4.5770 ms |  1.00 |    0.00 |  14000.0000 |           - |           - |          44065592 B |
         OptimizedXmlGetAll |  13.112 ms | 12.5939 ms | 0.6903 ms |  0.12 |    0.01 |   1750.0000 |           - |           - |           5849228 B |
        OptimizedDictGetAll |  16.207 ms | 15.3885 ms | 0.8435 ms |  0.15 |    0.01 |   1857.1429 |           - |           - |           5849229 B |
      OptimizedDictGetAllV2 |   1.305 ms |  0.7637 ms | 0.0419 ms |  0.01 |    0.00 |           - |           - |           - |                   - |
       OptimizedXmlLocalize |   7.173 ms |  5.8403 ms | 0.3201 ms |  0.07 |    0.01 |    100.0000 |           - |           - |            361272 B |
      OptimizedDictLocalize |   1.792 ms |  0.7922 ms | 0.0434 ms |  0.02 |    0.00 |    104.1667 |           - |           - |            360624 B |
// * Warnings *
MinIterationTime
  LocalizedTextServiceGetAllStoredValuesBenchmarks.OriginalDictionaryLocalize: IterationCount=3, IterationTime=100.0000 ms, LaunchCount=1, WarmupCount=3 -> MinIterationTime = 96.0692 ms which is very small. It's recommended to increase it.
  LocalizedTextServiceGetAllStoredValuesBenchmarks.OriginalXmlLocalize: IterationCount=3, IterationTime=100.0000 ms, LaunchCount=1, WarmupCount=3        -> MinIterationTime = 79.0338 ms which is very small. It's recommended to increase it.
  LocalizedTextServiceGetAllStoredValuesBenchmarks.OptimizedXmlGetAll: IterationCount=3, IterationTime=100.0000 ms, LaunchCount=1, WarmupCount=3         -> MinIterationTime = 95.6325 ms which is very small. It's recommended to increase it.
  LocalizedTextServiceGetAllStoredValuesBenchmarks.OptimizedDictGetAll: IterationCount=3, IterationTime=100.0000 ms, LaunchCount=1, WarmupCount=3        -> MinIterationTime = 85.7158 ms which is very small. It's recommended to increase it.

// * Legends *
  Mean                : Arithmetic mean of all measurements
  Error               : Half of 99.9% confidence interval
  StdDev              : Standard deviation of all measurements
  Ratio               : Mean of the ratio distribution ([Current]/[Baseline])
  Gen 0/1k Op         : GC Generation 0 collects per 1k Operations
  Gen 1/1k Op         : GC Generation 1 collects per 1k Operations
  Gen 2/1k Op         : GC Generation 2 collects per 1k Operations
  Allocated Memory/Op : Allocated memory per single operation (managed only, inclusive, 1KB = 1024B)
  1 ms                : 1 Millisecond (0.001 sec)

// * Diagnostic Output - MemoryDiagnoser *


// ***** BenchmarkRunner: End *****
Run time: 00:00:11 (11.6 sec), executed benchmarks: 8
```

How?
Replaced LINQ/xpath lookups for localized text with dictionaries.
Added a second Dictionary to look up localized text when there is no alias.
Split up the key into area and alias for calls to Localize(). Localize() was having to split into area and alias so this saves on allocations from string.Split() and string.Format to generate the key.
Created new ITextLocalizationService2 interface for localize() passing area and alias.
Updated calls to TextLocalizationService to use ITextLocalizationService2 where possible.
Reduced code/allocations by converting XML into dictionaries and operating on those.
Replaced use of string.Format with string.Concat where valid as string.concat is faster and no formatting was required.
Wrote benchmarks to validate.
Because the key is split into it's area and alias, the area part of the string is now interned seperately from the alias at compile time, and the alias appears frequently, this reduces the length of the strings, saving on allocations.
Add 
`  IDictionary<string, IDictionary<string, string>> GetAllStoredValuesByAreaAndAlias(CultureInfo culture);`
To ITextLocalizationService2 avoiding the unnecessary conversion from a to b to a again, zero allocation :)

How to test?
Existing tests pass.

---
_This item has been added to our backlog [AB#12568](https://dev.azure.com/umbraco/243e7927-03b2-44e2-908f-d4ac7ea5daaa/_workitems/edit/12568)_

### CODE REVIEW

File: src/Umbraco.Core/Services/Implement/LocalizedTextService.cs [hunk 1]
```diff
@@ -211,27 +243,21 @@ private string GetFromDictionarySource(CultureInfo culture, string area, string
...
-            var cultureSource = _dictionarySource[culture];
 
-            string found;
-            if (area.IsNullOrWhiteSpace())
+            string found = null;
+            if (string.IsNullOrWhiteSpace(area))
             {
-                found = cultureSource
-                    .SelectMany(x => x.Value)
-                    .Where(keyvals => keyvals.Key.InvariantEquals(key))
```
Comment (nzdev) [inline src/Umbraco.Core/Services/Implement/LocalizedTextService.cs:221]: Dictionary comparer is Invariant case insensitive now.

**Comment by Shazwazza** [comment]
Yes! 10000%

Only other great thing would be some additional unit tests if possible. If you don't have time we can see if someone else can help out with that. 

@nul800sebastiaan can we prioritize this one somehow for the next minor?

**Comment by nzdev** [comment]
No time at the moment for extra unit tests unfortunately.

**Comment by nathanwoulfe** [comment]
@Shazwazza if this is ok to go without the additional tests, I'll get on it today. 

Curious now too as to whether a similar change for the client-side localization would be worthwhile - there's similar string allocation happening as the localizationService.localize method accepts a formatted string as a param (`area_key`), which is then split down to the component parts. Might have a look and see if the benefit is worth the effort... If not, the way the current splitting is handled isn't as efficient as it could be.

**Comment by Shazwazza** [comment]
> if this is ok to go without the additional tests, I'll get on it today.

It should be but it still needs some manual testing ... and ideally we then add some extra test cases just for the new `ILocalizedTextService2` method (would be quite easy to do).

> Curious now too as to whether a similar change for the client-side localization would be worthwhile

most likely, allocations in the JS world would have the same issue as the c# world

**Comment by nzdev** [comment]
Doesn't pick up package text.

**Comment by nzdev** [comment]
Does pick up package text.

**Comment by Shazwazza** [comment]
Just reviewing now and fixing up a couple things. @nzdev you mentinoed

> the area part of the string is now interned seperately

but I'm not seeing any string interning in this PR?

**Comment by Shazwazza** [comment]
ah sorry i see what you mean - compile time. Yes because we are no longer splitting strings. Can't wait to have this released! 🎉 have marked as 8.15 (which is officially 8.CHAD)

**Comment by nul800sebastiaan** [comment]
I've made a PR to fix both issues above: https://github.com/umbraco/Umbraco-CMS/pull/10767

**Pull request merged**